### PR TITLE
refactor: extract Chainspec type from primitives 

### LINKED ISSUE #7929: extract Chainspec type from primitives
The chainspec type bloats reth-primitives

https://github.com/paradigmxyz/reth/blob/cc4a418ddf73ae5af240b403e4b6e1d191130cee/crates/primitives/src/chain/spec.rs#L509-L517

the entire mod should be converted into its own crate

https://github.com/paradigmxyz/reth/blob/cc4a418ddf73ae5af240b403e4b6e1d191130cee/crates/primitives/src/chain/mod.rs

this type is used quite a lot, we can likely re-export this type from reth-provider or similar

### PR DESCRIPTION
**Pull request by guha-rahul**
re: https://github.com/paradigmxyz/reth/issues/8674

Fixes: #7929 
The error i am facing -  
```
error: failed to load manifest for workspace member `/home/rahul/reth/bin/reth`

Caused by:
  failed to load manifest for dependency `reth-basic-payload-builder`

Caused by:
  failed to load manifest for dependency `reth-engine-primitives`

Caused by:
  failed to load manifest for dependency `reth-chainspec`

Caused by:
  failed to load manifest for dependency `reth-primitives`

Caused by:
  failed to parse manifest at `/home/rahul/reth/crates/primitives/Cargo.toml`

Caused by:
  error inheriting `reth_chainspec` from workspace root manifest's `workspace.dependencies.reth_chainspec`

Caused by:
  `dependency.reth_chainspec` was not found in `workspace.dependencies`
  ```
  Even tho i have `reth-chainspec = { path = "crates/storage/chainspec" }`  under `[workspace.dependencies]` .

### CODE REVIEW

File: crates/storage/chainspec/Cargo.toml
Comment (mattsse) [inline crates/storage/chainspec/Cargo.toml:1]: I want to move this to `crates/chainspec`

File: crates/storage/chainspec/Cargo.toml
Comment (mattsse) [inline crates/storage/chainspec/Cargo.toml:9]: needs update

File: crates/storage/chainspec/src/lib.rs
Comment (mattsse) [inline crates/storage/chainspec/src/lib.rs:1]: needs regular lib.rs header, see other crates for ref

File: crates/primitives/Cargo.toml
Comment (mattsse) [inline crates/primitives/Cargo.toml:21]: I believe this is your error

```suggestion
reth-chainspec.workspace = true
```

**Review by mattsse — CHANGES_REQUESTED** [review]
nice, a few nits and suggestions

File: crates/storage/chainspec/Cargo.toml
Comment (guha-rahul) [reply crates/storage/chainspec/Cargo.toml]: there are some extra imports, will remove them when i get back electricity.

File: crates/primitives/Cargo.toml
Comment (guha-rahul) [reply crates/primitives/Cargo.toml]: Got a cycle dependency error:- 
```
error: cyclic package dependency: package `reth-chainspec v0.2.0-beta.6 (/home/rahul/reth/crates/chainspec)` depends on itself. Cycle:
package `reth-chainspec v0.2.0-beta.6 (/home/rahul/reth/crates/chainspec)`
    ... which satisfies path dependency `reth-chainspec` (locked to 0.2.0-beta.6) of package `reth-primitives v0.2.0-beta.6 (/home/rahul/reth/crates/primitives)`
    ... which satisfies path dependency `reth-primitives` (locked to 0.2.0-beta.6) of package `reth-chainspec v0.2.0-beta.6 (/home/rahul/reth/crates/chainspec)`
    ... which satisfies path dependency `reth-chainspec` (locked to 0.2.0-beta.6) of package `reth-engine-primitives v0.2.0-beta.6 (/home/rahul/reth/crates/engine-primitives)`
    ... which satisfies path dependency `reth-engine-primitives` (locked to 0.2.0-beta.6) of package `reth-basic-payload-builder v0.2.0-beta.6 (/home/rahul/reth/crates/payload/basic)`
    ... which satisfies path dependency `reth-basic-payload-builder` (locked to 0.2.0-beta.6) of package `reth v0.2.0-beta.6 (/home/rahul/reth/bin/reth)`
    ... which satisfies path dependency `reth` (locked to 0.2.0-beta.6) of package `beacon-api-sse v0.0.0 (/home/rahul/reth/examples/beacon-api-sse)`
    ```

**Review by mattsse — CHANGES_REQUESTED** [review]
we also need to delete chainspec from primitives crate

**Comment by guha-rahul** [comment]
hey @mattsse , just a quick clarification , when i am removing `reth-chainspec` from primitives, there are some functions that are imported to other crates, eg: `revm_spec_by_timestamp_after_merge`, so for those we have to explicitly pass the 'reth-primitives' or there is something i am missing?

**Comment by mattsse** [comment]
these functions we should also move

**Comment by guha-rahul** [comment]
> these functions we should also move
I have identified 5 functions in the `reth-primitives` crate that are being used in other crates and they have chainspec in params.  
eg: `revm_spec_by_timestamp_after_merge` is defined in `primitive crate` and i am having a mismatch type in `reth-engine-primitives` crate.
Can you please elaborate more on how we should move the functions?

**Comment by mattsse** [comment]
any chainspec usage needs to be removed from reth-primitives

**Comment by guha-rahul** [comment]
When i am using `make pr `, this error occurs
```
error[E0412]: cannot find type `ChainSpec` in this scope
   --> crates/storage/provider/src/bundle_state/bundle_state_with_receipts.rs:180:22
    |
180 |         chain_spec: &ChainSpec,
    |                      ^^^^^^^^^ not found in this scope
    |
```

Although when i try to add  `use reth_primitives::ChainSpec;`, its showing me an unused import error. Any pointers on why that might be happening?

**Comment by guha-rahul** [comment]
hey @mattsse , any comments on the above commits?

**Review by mattsse — COMMENTED** [review]
hmm, need to take a closer look at all the deps, but in essence this is exactly what we want

**Review by mattsse — CHANGES_REQUESTED** [review]
okay, I made some progress here on figuring out how things should be structured.

what we need to do is keep a dependency on reth-primitives in the chainspec crate, so `chainspec depends on reth-primitives`

this should do it

**Comment by guha-rahul** [comment]
> okay, I made some progress here on figuring out how things should be structured.
> 
> what we need to do is keep a dependency on [reth-primitives](https://linear.app/paradigm-oss/issue/ID--primitives) in the chainspec crate, so `chainspec depends on reth-primitives`
> 
> this should do it

Do we need to have the Optimism mainnet spec?

**Review by mattsse — COMMENTED** [review]
this is great, ty.

need to think about how to phase out the primitives code, but this new crate is exactly what we want

**Comment by guha-rahul** [comment]
> this is great, ty.
> 
> need to think about how to phase out the primitives code, but this new crate is exactly what we want

Do we need to replace it with the alloy-primitives?

**Comment by Rjected** [comment]
bump @mattsse, this looks ok, the old json genesis assets can be deduplicated probably?

**Comment by guha-rahul** [comment]
Regarding the failed tests, i am struggling to understand why i am getting errors in `crates/net/eth-wire-types/src/receipt.rs`

**Review by Rjected — CHANGES_REQUESTED** [review]
could you remove the old chainspec now that it's in a separate crate, along with everything else that was copied to this new crate? we don't want to keep any duplicate files around

**Comment by mattsse** [comment]
I also want merge this after we have merged all changes in #8214

**Comment by guha-rahul** [comment]
> could you remove the old chainspec now that it's in a separate crate, along with everything else that was copied to this new crate? we don't want to keep any duplicate files around

hey when I am trying to import the new chainspec in place of the old one , I am getting cyclic dependency errors and @mattsse previously mentioned we cannot import chainspec in reth-primitives since it will cause a cyclic dependency error.

**Review by mattsse — APPROVED** [review]
lgtm

**Review by Rjected — APPROVED** [review]
lgtm

**Pull request merged**
